spinoso-string/src/enc/utf8/mod.rs

```diff
@@ -0,0 +1,1060 @@
+                    // ```
+                    // substr = "a\xF0\x9F\x87"
+                    // slice = "\xF0\x9F\x87"
+                    // remaining = 1
+                    // ```
+                    //
+                    // We want to compute `endth == 2`:
+                    //
+                    //    2   =      4       -      3      +     1
+                    let endth = substr.len() - slice.len() + remaining;
+                    return Some(&substr[..endth]);
+                }
+                // If there are more characters remaining than the number
+                // of bytes yielded in the invalid UTF-8 byte sequence,
+                // count `size` bytes and advance the slice to continue
+                // decoding.
+                (None, size) => {
+                    slice = &slice[size..];
+                    remaining -= size;
+                }
+            }
+        }
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn get_mut<I>(&mut self, index: I) -> Option<&mut I::Output>
+    where
+        I: SliceIndex<[u8]>,
+    {
+        self.inner.get_mut(index)
+    }
+
+    #[inline]
+    #[must_use]
+    pub unsafe fn get_unchecked<I>(&self, index: I) -> &I::Output
+    where
+        I: SliceIndex<[u8]>,
+    {
+        self.inner.get_unchecked(index)
+    }
+
+    #[inline]
+    #[must_use]
+    pub unsafe fn get_unchecked_mut<I>(&mut self, index: I) -> &mut I::Output
+    where
+        I: SliceIndex<[u8]>,
+    {
+        self.inner.get_unchecked_mut(index)
+    }
+}
+
+// Pushing and popping bytes, codepoints, and strings.
+impl Utf8String {
+    #[inline]
+    pub fn push_byte(&mut self, byte: u8) {
+        self.inner.push_byte(byte);
+    }
+
+    #[inline]
+    pub fn try_push_codepoint(&mut self, codepoint: i64) -> Result<(), InvalidCodepointError> {
+        let codepoint = if let Ok(codepoint) = u32::try_from(codepoint) {
+            codepoint
+        } else {
+            return Err(InvalidCodepointError::codepoint_out_of_range(codepoint));
+        };
+        if let Ok(ch) = char::try_from(codepoint) {
+            self.push_char(ch);
+            Ok(())
+        } else {
+            Err(InvalidCodepointError::invalid_utf8_codepoint(codepoint))
+        }
+    }
+
+    #[inline]
+    pub fn push_char(&mut self, ch: char) {
+        self.inner.push_char(ch);
+    }
+
+    #[inline]
+    pub fn push_str(&mut self, s: &str) {
+        self.inner.push_str(s);
+    }
+
+    #[inline]
+    pub fn extend_from_slice(&mut self, other: &[u8]) {
+        self.inner.extend_from_slice(other);
+    }
+}
+
+// Encoding
+impl Utf8String {
+    #[inline]
+    #[must_use]
+    pub fn is_ascii_only(&self) -> bool {
+        self.inner.is_ascii()
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn is_valid_encoding(&self) -> bool {
+        if self.is_ascii_only() {
+            return true;
+        }
+
+        simdutf8::basic::from_utf8(&self.inner).is_ok()
+    }
+}
+
+// Casing
+impl Utf8String {
```

Comment:
Use `roe` for case changing operation in `spinoso-string`.

UTF-8 case changing needs to be parameterized on the casefolding strategy to account for e.g. Turkic or ASCII-only modes.

Reply:
(no action needed in this PR. I left this comment to make it easier to file a ticket.)

Reply:
Added as a TODO too

Reply:
File:

- https://github.com/artichoke/artichoke/issues/1723